packages/flutter_tools/lib/src/macos/application_package.dart

```diff
@@ -58,8 +63,26 @@ abstract class MacOSApp extends ApplicationPackage {
       }
       bundleDir = globals.fs.directory(applicationBundle);
     } else {
-      globals.printError('Folder "${applicationBundle.path}" is not an app bundle.');
-      return null;
+      // Try to unpack as a zip.
+      final Directory tempDir = globals.fs.systemTempDirectory.createTempSync('flutter_app.');
+      shutdownHooks.addShutdownHook(() async {
+        await tempDir.delete(recursive: true);
+      }, ShutdownStage.STILL_RECORDING);
+      try {
+        globals.os.unzip(globals.fs.file(applicationBundle), tempDir);
+      } on ProcessException {
+        globals.printError('Invalid prebuilt macOS app. Unable to extract bundle from archive.');
+        return null;
+      }
+      try {
+        bundleDir = tempDir
```

Comment:
rather than catch the state error, I would use the firstWhere orElse:() => null pattern

Reply:
firstWhere handles the case of no application bundle but it does not handle the case of multiple bundles in the zip file. To handle both without the StateError we could check that the length of the iterable returned by where(_isBundleDirectory) is equal to one. But using singleWhere we actually get a StateError thrown for both of those cases and it seems like this actually looks a bit nicer.

Reply:
Yeah, that is a good point. SGTM